man/radlast.1.man
.\" Copyright (c) 2001 Francesco Paolo Lovergine <francesco@yardradius.org>
.\" See copyright section for conditions of redistribution
.\" $Id: radlast.1.man,v 1.4 2004/07/01 13:57:55 flovergine Exp $
.\" 
.TH "RADLAST" "1" "Jul 1, 2004" "1.0.19" "Yard Radius Manual"
.SH "NAME"
radlast \- lists last users access and stats 
.SH "SYNOPSIS"
.BR radlast " [ "
.BR "\-chbx" " ] ["
.B "\-a "
.I R acct_dir " ] ["
.B "\-f "
.I R file " ] ["
.B "\-F "
.I R format_string " ] ["
.B "\-H "
.I R format_string " ] ["
.B "\-m " 
.I R month " ] ["
.B "\-u "
.I R user " ] ["
.B "\-y "
.I R year " ]"
.SH "DESCRIPTION"
This program gives a detailed list of last accesses for one or more
users. It does not collect statistical information. It's a fast
and compact dumper of the traditional `detail' files of Livingston's
Radius. It's output is in some way similar to that of the Unix command `last'
but also high\-customizable (see FORMATS section below).
.SH "OPTIONS"
.TP 
.BR "\-a " acct_dir
Sets the accounting files directory prefix, instead of the 
default one, which is 
.I PREFIX/logs
.TP 
.B \-b
Uses a single\-row brief output format. 
.TP 
.B \-c
Shows records in cronological order.
.TP 
.BR "\-f " file
Uses this radlast logfile instead of the default stdout.
.TP 
.B \-h
Prints out usage of the command.
.TP 
.BR "\-F \-H " format_string
Define customized formats for output rows and the header. 
See FORMATS section below.
.TP 
.BR "\-m " month
Reports statistics for month 
.I month
where month is in the range 1\-12.
.TP 
.BR "\-n " num
Shows no more than num entries.
.TP 
.BR "\-y " year
Reports statistics for year
.I year
that is a four digits number.
.TP 
.B \-x
Shows extedend information (include traffic and phone CLIs)
.P
.SH "FORMATS"
.B radlast
does allow the user to  have a customized output format with the `\-F' 
command line option (see above). The program accepts and outputs any
character in a format string and parse it to find some %\-prefixed
one\-letter tokens. Those `a la' printf tokens are substituted with
corresponding entry values.
Normal C substitution are performed for \\\\,\\t,\\n,\\r and \\%.
The list of legal % tokens follows, any oher combination is parsed 
and ignored.

.TS
l l.
Token	Value
_
%l	Username
%p	Port ID
%a	NAS ip address
%k	Port type
%c	Client ip address
%d	Date in ctime format
%t	Online time in secs
%T	Online time in HH:MM:SS format
%i	Input traffic in bytes
%I	Input traffic in KBytes
%o	Output traffic in bytes
%O	Output traffic in KBytes
%m	Total traffic in bytes
%M	Total traffic in KBytes
%s	Input speed in bps (or UNKNOWN)
%S	Output speed in bps (or UNKNOWN)
%A	NAS called id (or UNKNOWN)
%#	Client calling id (or UNKNOWN)
%%	literal '%'
.TE

.SH "FILES"
.TP 
.I "PREFIX/logs/YEAR/radlast\-XX"
These are the files which constitutes the radlast database. They are in 
binary not\-indexed format and functionally equivalent to the text files
`detail\-XX'. The main difference is that detail files are organized on
a per client base. See below.
.TP 
.I "PREFIX/logs/NAS/YEAR/detail\-XX"
These text files stores accounting information for every access server
(NAS) listed in the 
.I clients
file. Each user session is identified by a unique session\-id and these
accounting files store start/stop records for every single authenticated
session. Information stored in those records, but for a group of standard
ones, strictly depend on the NAS model and operating system. 
.SH "BUGS"
The 
.BR "\-x " option
apparently is not working.
.SH "SEE ALSO"
.BR builddbm (8),
.BR radiusd (8),
.BR radlist (1),
.BR radtest (1),
.BR radwatch (1),
.BR radius_attributes (5),
.SH "AUTHOR"
Francesco Paolo Lovergine <francesco@yardradius.org>.
.P
A complete list of contributors is contained in CREDITS file. 
You should get that file among other ones within your distribution 
and possibly installed under
.BR "PREFIX/docs " directory
.SH "COPYRIGHT"
Copyright (C) 1997 Cristian Gafton. All rights reserved.
.P
Copyright (C) 1999\-2001 Francesco Paolo Lovergine. All rights reserved.
.P
See the LICENSE file enclosed within this software for conditions of 
use and distribution. This is a pure 
.B "ISO BSD Open Source License" .


ChangeLog
[ $Id: ChangeLog,v 1.29 2004/07/01 13:57:52 flovergine Exp $ ]

(1.0.21) 2004-07-01 Francesco P. Lovergine <francesco@yardradius.org>

	The 'middle-land' release.

	  [Minor]
	 
	* Editorial changes in copyright notices of all source files.
	  Also some typos are now corrected.
	* New email and domain used extensively in every file.
	* Updated README to reflect new site URL and other info.
	* Refinement in size definitions of some data structures.

	  [Enhancement]
	  
	* Changed --version in -dumpversion for gcc versioning 
          in configure.in. This broke things in gcc 3. New arg is
          back compatible also.
	* Added entries limit to radlast. Thanks Russell Heilling 
	  for patch (see feature request 464791). Changed man page
	  as consequence.
	* Updated config.guess and config.sub, old releases could
	  break thing on current systems.

	  [Security]

	* A buffer overflow in calc_acctreq() has been patched.
	* process_menu() has been modified to check string limits.
	  Thanks Max Vozeler for pointing this so old issue and
	  showed a possible exploit.

(1.0.20) 2002-03-02 Francesco P. Lovergine <francesco@yardradius.org>
	* Documentation has been revised and updated. Many typos are
	  now corrected as well as some documentation incoherences.
	  New RFCs about the RADIUS protocol suite are enclosed and
	  obsolete ones removed.
	* Some files changed names for coherence.
	* getopt() returns int, not char.
	* SIGABRT is the standard definition in POSIX.1 which obsoletes
	  SIGIOT. In some environment the old signal is no more defined.
	* Now changeprefix script works good on Free BSD. The non
	  portable '-' argument has been skipped for sed. This caused
	  problem with configure script under that platform.
	* Dictionary updated with other VSAs (Ericsson and Xedia).
	* config.sub, config.guess and other automake scripts are now 
	  updated to the latest releases.
	* Major code revisions to avoid vulnerabilities due to strcpy()
          memcpy() and other dangerous functions. 
           - Digest Calculation Buffer Overflow Vulnerability.
           - Invalid attribute length calculation on malformed VSA.
          Non full-disclosures are available at CERT.org
	* Added my GPG key to code.

	  Some of those changes are stolen from a VA Linux patch for
	  Lucent Radius 2.1.

(1.0.19) 2001-08-27 Francesco P. Lovergine <francesco@yardradius.org>
	* Added a first version of man files under `man' directory.
	  Some old draft documentation files have been removed,
	  as consequence.
	* Default prefix is changed in `/usr/local/yardradius'
	* A couple of commas in configure.in which caused a ',6' 
	  file when --enable-debian was used has been deleted. 
	  A minor, but truly amazing bug :)
	* Many typos and grammar errors in doc corrected (or hoping so).
	* Configure.in did not install logs and doc directory when
	  prefix is not set on command line. Now also this is ok.

(1.0.18) 2001-08-21 Francesco P. Lovergine <francesco@yardradius.org>
	* Patched for a subdle bug due to __debug definition in extens.c.
	* Skipped auth_item check when Auth-Type=PAM.
	* Skipping void reply items in parsing GDBM users files.

	  All patches are due to William K. Volkman.

(1.0.17) 2001-08-18 Francesco P. Lovergine <francesco@yardradius.org> 
	* Configuration file modified to add a --enable-debian option.
	  It greatly simplifies support of Debian packaging. Debian
	  package is released as a separate package in Debian 'unstable'
	  distribution.
	* Now builddbm and radiusd are installend under sbin which
	  seems a more correct destination for those files.

(1.0.16/pre16) 2001-01-14 Francesco P. Lovergine <francesco@yardradius.org> 
	* Dictionary entries for BayNetworks/Nortel Annex is added.
	* Now, YARD RADIUS is compatible with GDBM 1.7.3, which seems the 
	  most commonly installed version. Just changed the gdbm_open()
	  function arguments for this.
	* Since now, the official name of the package is
	  Yard Radius. This to avoid confusion with other packages
	  with the same name, and to be coherent with a next-to-come 
	  Debian package (yeah!).
	  I changed also version numbering. The current version is 
	  1.0.16, i.e. 1.0 patch level 16. It seems more `standard',
	  and to be honest I really think this program is too much
	  robust to leave it in pre-release status... Am I an optimist?
	* Changed including of time.h and sys/time.h as suggested in autoconf
	  manuals. This allows installation under AIX and other systems
	  which could need both files. Incidentally, under AIX it compiles
	  also with the off-the-shelf C compiler.
	* PAM configuring in configure.in is changed to avoid a 
	  subdle configuration bug: some linux distributions do not
	  install PAM development libraries, so that PAM support cannot
	  be enabled on them, while building. This includes Debian 2.2
	  and RedHat.  That's not a problem on Solaris or FreeBSD 
	  which install them as default. This require a modification
	  of configure.in which is now smarter. 
	* An silly ortographic error in configure.in with --enable-pam
	  has been corrected.
	* Builddbm now works under the conf directory, not the current one.
	* License and Readme files have been changed. There is no more
	  an ANNOUNCE file and COPYING (which is the standard name for
	  GPL) is now named LICENSE (which is the standard name for BSD).
	* A CONFIGURATION text file has been added to summarize main
	  configuration issues. Are you happy?
	* Many other (little) adjustments.
	
(pre15) 2000-09-26 Francesco P. Lovergine <francesco@yardradius.org>
	* Installer now create also logs directory and explains
	  the structure.
	* Deleted Log entries from sources to avoid redundant comments
	  with CVS.

(pre14) 2000-09-21 Francesco P. Lovergine <francesco@yardradius.org>
	* Corrected autoconf/automake 
	  configuration to solve some problems of installation
	  under a few architectures. Configure now use Gnu Shell Tool

src/radlast.c
/*
 * Copyright (C) 1999-2002 Francesco P. Lovergine. 
 * All rights reserved.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms stated in the LICENSE file which should be
 * enclosed with sources.
 */

static char rcsid[] = "$Id: radlast.c,v 1.8 2004/07/01 13:57:55 flovergine Exp $";

# define __MAIN__

#include	"yard.h"
#include	"global.h"

char *progname = NULL;
char username[USERNAME_MAX];

static int brief = 0;
static int extended = 0;

static const char *port_type_str(int type)
{
    switch (type) {
	case 0:
	    return "ASYN";
	case 1:
	    return "SYNC";
	case 2:
	    return "ISDN";
	case 3:
	    return "V120";
	case 4:
	    return "V110";
	default:
	    return "UNKN";
    }
    /* not reached */
    return "Unknown";
}

static const char * proto_type_str(int proto)
{
    switch (proto) {
	case P_TELNET:
	    return "TELN";
	case P_RLOGIN:
	    return "RLOG";
	case P_TCP_CLEAR:
	    return "DATA";
	case P_PORTMASTER:
	    return "PM";
	case P_PPP:
	    return "PPP";
	case P_SLIP:
	    return "SLIP";
	case P_CSLIP:
	    return "CSLP";
	default:
	    return "UNK";
    }
    /* not reached */
    return "UNK";
}


static int parse_width(char **format)
{
  int width=0;
  char *endstr;

  if ( isdigit(**format) || **format=='+' || **format=='-' )
    {
    width=(int)strtol(*format,&endstr,10);
    *format=endstr;
    }
  return width;
}

/*
 *	Function: parse_formatstr()
 * 	Purpose: parse a -F format string and outputs as required all
 *		 radlast fields.
 *
 *	Format tokens are:
 *		%l	Username
 *		%p	Port ID
 *		%a	NAS ip address
 *		%k	Port type
 *		%c	Client ip address
 *		%d	Date in ctime format
 *		%t	Online time in secs
 *		%T	Online time in HH:MM:SS format
 *		%i	Input traffic in bytes
 *		%I	Input traffic in KBytes
 *		%o	Output traffic in bytes
 *		%O	Output traffic in KBytes
 *		%m	Total traffic in bytes
 *		%M	Total traffic in KBytes
 *		%s	Input speed in bps (or UNKNOWN)
 *		%S	Output speed in bps (or UNKNOWN)
 *		%A	NAS called id (or UNKNOWN)
 *		%#	Client calling id (or UNKNOWN)
 *		\n \t \\ \% %% As obvious :-)
 *
 *	Tokens can include an integer signed value for alignment with
 * 	the same sense of printf() format value.
 */

static void parse_formatstr( char *formatstr,radlast rl )
{
  char port_str[100];
  char nas_ip[100];
  char client_ip[100];
  char fstr[100];
  char *str_time;
  int  width;
  char *ptr;
  time_t login_time;


  if ( formatstr == NULL ) return;

  memset(port_str, 0, sizeof(port_str));
  memset(nas_ip, 0, sizeof(nas_ip));
  memset(client_ip, 0, sizeof(client_ip));
  login_time = rl.ut_time;
  str_time = ctime(&login_time); str_time[19] = '\0';

  snprintf(port_str, 
	   sizeof(port_str), 
           "%4.4s-%02d",
           port_type_str(rl.ent.port_type),
           rl.ent.port);
  ipaddr2str(nas_ip, sizeof(nas_ip), rl.nas_ip);
  ipaddr2str(client_ip, sizeof(client_ip), rl.client_ip);

  while ( *formatstr )
    {
    switch ( *formatstr )
      {
      case '%':
        formatstr++;
        width=parse_width(&formatstr);
        if (width) snprintf(fstr,sizeof(fstr),"%%%d",width);
        else strcpy(fstr,"%");
	switch (*formatstr)
	  {
	  case 'l':
	    strcat(fstr,"s"); printf(fstr,rl.login); 
	    break;
	  case 'p':
	    strcat(fstr,"s"); printf(fstr,port_str); 
	    break;
	  case 'a':
	    strcat(fstr,"s"); printf(fstr,nas_ip); 
	    break;
	  case 'k':
	    strcat(fstr,"s"); printf(fstr,proto_type_str(rl.ent.proto)); 
	    break;
	  case 'c':
	    strcat(fstr,"s"); printf(fstr,client_ip); 
	    break;
	  case 'd':
	    strcat(fstr,"s"); printf(fstr,str_time); 
	    break;
	  case 't':
	    strcat(fstr,"ld"); printf(fstr,rl.length); 
	    break;
	  case 'T':
	    ptr=fstr+50;
	    snprintf(ptr,sizeof(fstr)/2,"%dh:%dm:%ds",
		   (int)(rl.length/3600),
                   (int)((rl.length%3600)/60),
                   (int)((rl.length%3600)%60)); 
	    strcat(fstr,"s");
	    printf(fstr,ptr);
	    break;
	  case 'i':
	    strcat(fstr,"ld"); printf(fstr,rl.inb); 
	    break;
	  case 'I':
	    strcat(fstr,"ldKB"); printf(fstr,rl.inb/1024); 
	    break;
	  case 'o':
	    strcat(fstr,"ld"); printf(fstr,rl.outb); 
	    break;
	  case 'O':
	    strcat(fstr,"ldKB"); printf(fstr,rl.outb/1024); 
	    break;
	  case 'm':
	    strcat(fstr,"ld"); printf(fstr,rl.outb+rl.inb); 
	    break;
	  case 'M':
	    strcat(fstr,"ldKB"); printf(fstr,(rl.outb+rl.inb)/1024); 
	    break;
	  case 's':
	    if (rl.rxrate) {
	      strcat(fstr,"d"); printf(fstr,rl.rxrate); 
	    }
	    else {
	      strcat(fstr,"s"); printf(fstr,"NONE");
	    }
	    break;
	  case 'S':
	    if (rl.txrate) {
	      strcat(fstr,"d"); printf(fstr,rl.txrate); 
	    }
	    else {
	      strcat(fstr,"s"); printf(fstr,"NONE");
	    }
	    break;
	  case 'A':
	    strcat(fstr,"s"); 
	    printf(fstr,strlen(rl.calledid)?rl.calledid:"NONE"); 
	    break;
	  case '#':
	    strcat(fstr,"s"); 
	    printf(fstr,strlen(rl.callingid)?rl.callingid:"NONE"); 
	    break;
	  case '%':
	    putchar('%');
	  default:
	    break;
	  }
        break;
      case '\\':
        formatstr++;
        switch(*formatstr)
	  {
	  case 'n':
	    printf("\n");
	    break;
	  case 't':
	    printf("\t");
	    break;
	  case 'r':
	    printf("\r");
	    break;
	  case '\\':
	  default:
	    printf("\\");
	    break;
	  }
        break;
      default:
	putchar(*formatstr);
	break;
      }
    formatstr++;
    }
}

/*
 *	Function: parse_header()
 * 	Purpose: parse a -H header string 
 *
 *	Format tokens are:
 *		\n \t \r \\ As obvious :-)
 */

static void parse_header( char *header )
{
   while (*header) {
      switch (*header) {
      case '\\':
        header++;
        switch(*header)
	  {
	  case 'n':
	    printf("\n");
	    break;
	  case 't':
	    printf("\t");
	    break;
	  case 'r':
	    printf("\r");
	    break;
	  case '\\':
	  default:
	    printf("\\");
	    break;
	  }
        break;
      default:
	putchar(*header);
	break;
      }
    header++;
    }
}

static void print_entry(radlast rad_last)
{
  if (extended)
    if (brief) parse_formatstr("%l/%p/%a/%k/%c/%d/%T/%I/%O/%M/%s/%S/%A/%#\n",
                          rad_last);
    else parse_formatstr( 
		    "\nUsername = %l\n"
	    	    "Port-Number = %p\n"
		    "NAS-IP = %a\n"
		    "Service-Type = %k\n"
		    "User-IP = %c\n"
		    "Date = %d\n"
		    "Session-Time = %T\n"
		    "Input-Traffic = %I\n"
		    "Output-Traffic = %O\n"
		    "Total-Traffic = %M\n"
		    "Rx-Speed = %s\n"
		    "Tx-Speed = %S\n"
		    "NAS-CLI = %A\n"
		    "User-CLI = %#\n",rad_last );
  else /* !extended */
	if (brief)
	    parse_formatstr("%l/%p/%a/%k/%c/%d/%T\n",rad_last);
	else
	    parse_formatstr( 
                    "\nUsername = %l\n"
                    "Port-Number = %p\n"
                    "NAS-IP = %a\n"
                    "Service-Type = %k\n"
                    "User-IP = %c\n"
                    "Date = %d\n"
                    "Session-Time = %T\n",rad_last);
}

static void radlast_usage(void)
{
    printf("\nUsage: %s <option>\n\nValid options are:\n", progname);
    printf("\t-a <acct_dir>\tthe radius accounting files directory\n");
    printf("\t-F <format>\tcustom output format\n");
    printf("\t-H <format>\tcustom header\n");
    printf("\t-c\t\tshow records in cronological order\n");
    printf("\t-u <username>\tget username database entry\n");
    printf("\t-f <filename>\tuse this radlast logfile instead\n");
    printf("\t-h\t\tthis help screen\n");
    printf("\t-n <num>\tPrint only <num> entries\n");
    printf("\t-m <month>\treport statistics for <month>\n");
    printf("\t-y <year>\treport statistics for <year>\n");
    printf("\t-b\t\tsingle row output\n");
    printf("\t-x\t\textended info (include traffic and CLIs)\n\n");
    exit(-1);
}

static void fatal_func(const char *msg) {
    fprintf(stderr, "FATAL ERROR: %s\n",msg);
    return;
}

int main(int argc, char **argv)
{
    radlast rad_last;
    int fd;
    static char filename[PATH_MAX];
    int flag;
    int month;
    int year;
    time_t crt_time;
    struct tm *time_info;
    int file_arg = 0;
    int month_arg = 0;
    int year_arg = 0;
    int user_arg = 0;
    const char *radacct_dir;
    int cronologic = 0;
    char *format=NULL;
    char *header=NULL;
    extern char *optarg;
    int num = 0, count = 0;
    
    crt_time = time(NULL);
    time_info = localtime(&crt_time);
    month = time_info->tm_mon + 1;				    
    year = 1900+time_info->tm_year;
    memset(&rad_last, 0, sizeof(radlast));
    memset(filename, 0, sizeof(filename));
    memset(username, 0, sizeof(username));
    
    progname = argv[0];
    radacct_dir = RADACCT_DIR;
    
    while ((flag = getopt(argc, argv, "a:F:H:cu:f:by:m:xhn:")) != EOF) {
	char *cp;
	switch (flag) {
	    case 'a':
		radacct_dir = optarg;
		break;
	    case 'F':
		format = optarg;
		break;
	    case 'H':
		header = optarg;
		break;
	    case 'c':
		cronologic++;
		break;
	    case 'u':
		strncpy(username, optarg, USERNAME_MAX);
		user_arg++;
		break;
	    case 'f':
		if (month_arg || year_arg) {
		    fatal_func("you cannot specify a filename and a month number");
		    radlast_usage();
		}
		strncpy(filename, optarg, PATH_MAX);
		file_arg++;
		break;
            case 'b':
		brief++;
		break;
            case 'n':
                num = strtol(optarg, &cp, 10);
		if (*cp) {
                   fatal_func("number of entries must be an integer\n");
                   radlast_usage();
		}
                break;
	    case 'y':
		if (file_arg) {
                    fatal_func("you can not specify a year and a filename");
		    radlast_usage();
		}
		year = strtol(optarg, &cp, 10);
		if (*cp) {
		    fatal_func("year should be a number\n");
		    radlast_usage();
		}
		if (year < 1970) {
		    fatal_func("year should be greater than 1970\n");
		    radlast_usage();
		}
		year_arg++;
		break;
	    case 'm':
		if (file_arg) {
                    fatal_func("you can not specify a month number and a filename");
		    radlast_usage();
		}
		month = strtol(optarg, &cp, 10);
		if (*cp) {
		    /* not a number ... */
		    fatal_func("month should be a number\n");
		    radlast_usage();
		}
		if ((month < 1) || (month > 12)) {
		    fatal_func("month should be between 1 and 12\n");
		    /* not a valid month */
		    radlast_usage();
		}
		month_arg++;
		break;
	    case 'x':
		extended++;
		break;
	    case 'h':
	    default:
		radlast_usage();
	}
    }

    if (!file_arg)
	snprintf(filename, sizeof(filename),"%s/%d/%s-%02d", 
	        radacct_dir, year, RADIUS_LAST, month);
    
    if ((fd = open(filename, O_RDONLY)) >= 0) {
	off_t seekpos = 0;
	
	if (!cronologic)
	    seekpos = lseek(fd, -sizeof(radlast), SEEK_END);
	if ( format != NULL && header != NULL ) parse_header(header); 
	while ((seekpos >= 0) &&
	       read(fd, &rad_last, sizeof(radlast)) == sizeof(radlast) &&
	       (num == 0 || count < num)) {
	    if (user_arg && strncmp(username, rad_last.login, USERNAME_MAX)) {
		memset(&rad_last, 0, sizeof(rad_last));
		if (!cronologic)
		    seekpos = lseek(fd, -2*sizeof(radlast), SEEK_CUR);
		continue;
	    }
	    if ( format != NULL ) parse_formatstr(format,rad_last);
	    else print_entry(rad_last);
	    memset(&rad_last, 0, sizeof(rad_last));
	    if (!cronologic)
		seekpos = lseek(fd, -2*sizeof(radlast), SEEK_CUR);
            if (num)
		count++;
	}
    } else {
	fprintf(stderr, "Cannot open the radlast log file: %s\n",filename);
	exit(-1);
    }
    exit(0);
}

void rad_exit(int code)
{
 	exit(code);
}
